blob/blob.go

```diff
@@ -139,6 +139,41 @@ func (r *Reader) As(i interface{}) bool {
 	return r.r.As(i)
 }
 
+// WriteTo reads from r and writes to w until there's no more data or
+// an error occurs.
+// The return value is the number of bytes written to w.
+//
+// It implements the io.WriterTo interface.
+func (r *Reader) WriteTo(w io.Writer) (int64, error) {
+	_, nw, err := readFromWriteTo(r, w)
+	return nw, err
+}
+
+// readFromWriteTo is a helper for ReadFrom and WriteTo.
+// It reads data from r and writes to w, until EOF or a read/write error.
+// It returns the number of bytes read from r and the number of bytes
+// written to w.
+func readFromWriteTo(r io.Reader, w io.Writer) (int64, int64, error) {
+	buf := make([]byte, 1024)
+	var nr, nw int64
+	for {
+		nb, err := r.Read(buf)
+		nr += int64(nb)
+		if err == io.EOF {
```

Comment:
I think it's OK for `Read` to return `nb>0` and `err == io.EOF`. Are you sure this is handled correctly here?

From the docs:

"Callers should always process the n > 0 bytes returned before considering the error err. Doing so correctly handles I/O errors that happen after reading some bytes and also both of the allowed EOF behaviors. "

Reply:
Good catch. A conformance test actually caught this as well. Fixed.